File: bitcache/partitioned.py
"""Partition-aware staged retrieval.

Clusters binary codes into P partitions at build time. At search time,
routes the query to the top-R most relevant partitions and scans only
those. Reduces O(n) to O(n * R/P) per query.

Architecture:
  Build:
    1. Binary quantize all vectors
    2. Cluster binary codes into P partitions (k-means on binary centroids)
    3. Store partition assignments and centroids

  Search:
    1. Binary quantize query
    2. Compute Hamming distance to all P partition centroids
    3. Select top-R partitions (routing)
    4. Scan only vectors in those R partitions (binary Hamming)
    5. Float rerank top rf*k candidates
    6. Return top-k
"""

import numpy as np
from typing import Optional

from bitcache.quantize import quantize
from bitcache.search import hamming_distance_single


class PartitionedIndex:
    """Partition-aware two-stage retrieval.

    Clusters vectors into partitions for sublinear search. Only scans
    vectors in the most relevant partitions, then reranks with float.

    Args:
        dim: Vector dimensionality.
        n_partitions: Number of partitions (P). More = faster search, risk of missing neighbors.
        n_probe: Number of partitions to scan per query (R). More = better recall, slower.
        rerank_factor: Candidates from scanned partitions to rerank with float.
    """

    def __init__(
        self,
        dim: int,
        n_partitions: int = 32,
        n_probe: int = 4,
        rerank_factor: int = 100,
    ):
        if dim <= 0:
            raise ValueError("dim must be positive")
        self.dim = dim
        self.n_partitions = n_partitions
        self.n_probe = n_probe
        self.rerank_factor = rerank_factor
        self.n_bytes = (dim + 7) // 8

        self._codes: Optional[np.ndarray] = None
        self._vectors: Optional[np.ndarray] = None
        self._n_vectors: int = 0

        # Partition structures
        self._centroids: Optional[np.ndarray] = None  # (P, n_bytes) binary centroids
        self._assignments: Optional[np.ndarray] = None  # (n,) partition ID per vector
        self._partition_indices: Optional[list] = None  # partition_id -> list of vector indices

    def build(self, vectors: np.ndarray) -> None:
        """Build the partitioned index.

        Args:
            vectors: Float array of shape (n, dim).
        """
        vectors = np.asarray(vectors, dtype=np.float32)
        if vectors.shape[1] != self.dim:
            raise ValueError(f"expected dim={self.dim}, got {vectors.shape[1]}")

        n = len(vectors)
        self._n_vectors = n

        # Normalize and store
        norms = np.linalg.norm(vectors, axis=1, keepdims=True)
        norms = np.where(norms > 1e-10, norms, 1.0)
        self._vectors = vectors / norms

        # Binary quantize
        self._codes = quantize(self._vectors)

        # Cluster binary codes into partitions using binary k-means
        self._centroids, self._assignments = self._binary_kmeans(
            self._codes, self.n_partitions
        )

        # Build partition index (which vectors belong to each partition)
        self._partition_indices = [[] for _ in range(self.n_partitions)]
        for i, pid in enumerate(self._assignments):
            self._partition_indices[pid].append(i)

    def search(self, query: np.ndarray, k: int = 10) -> tuple[np.ndarray, np.ndarray]:
        """Partition-routed search.

        Args:
            query: Float vector of shape (dim,).
            k: Number of results.

        Returns:
            Tuple of (scores, indices). Scores are float32 inner products.
        """
        if self._codes is None or self._n_vectors == 0:
            return np.array([], dtype=np.float32), np.array([], dtype=np.int64)

        query = np.asarray(query, dtype=np.float32).ravel()
        norm = np.linalg.norm(query)
        query_unit = query / norm if norm > 1e-10 else query

        # Binary quantize query
        query_code = quantize(query_unit[None, :])[0]

        # Route: find closest partitions
        centroid_dists = hamming_distance_single(query_code, self._centroids)
        probe_partitions = np.argsort(centroid_dists)[:self.n_probe]

        # Gather candidate indices from probed partitions
        candidate_indices = []
        for pid in probe_partitions:
            candidate_indices.extend(self._partition_indices[pid])
        candidate_indices = np.array(candidate_indices, dtype=np.int64)

        if len(candidate_indices) == 0:
            return np.array([], dtype=np.float32), np.array([], dtype=np.int64)

        # Binary scan within probed partitions
        candidate_codes = self._codes[candidate_indices]
        hamming_dists = hamming_distance_single(query_code, candidate_codes)

        # Select top rf*k from candidates
        n_rerank = min(k * self.rerank_factor, len(candidate_indices))
        if n_rerank < len(candidate_indices):
            top_local = np.argpartition(hamming_dists, n_rerank)[:n_rerank]
        else:
            top_local = np.arange(len(candidate_indices))

        rerank_indices = candidate_indices[top_local]

        # Float rerank
        rerank_vectors = self._vectors[rerank_indices]
        scores = rerank_vectors @ query_unit

        # Top-k
        top_k_local = np.argsort(scores)[::-1][:k]
        final_indices = rerank_indices[top_k_local]
        final_scores = scores[top_k_local]

        return final_scores, final_indices

    def search_batch(self, queries: np.ndarray, k: int = 10) -> tuple[np.ndarray, np.ndarray]:
        """Batch search.

        Args:
            queries: Float array of shape (nq, dim).
            k: Number of results per query.

        Returns:
            Tuple of (scores, indices) each shape (nq, k).
        """
        queries = np.asarray(queries, dtype=np.float32)
        if queries.ndim == 1:
            queries = queries[None, :]

        nq = len(queries)
        all_scores = np.zeros((nq, k), dtype=np.float32)
        all_indices = np.zeros((nq, k), dtype=np.int64)

        for i in range(nq):
            scores, indices = self.search(queries[i], k=k)
            n = len(scores)
            all_scores[i, :n] = scores
            all_indices[i, :n] = indices

        return all_scores, all_indices

    def _binary_kmeans(
        self, codes: np.ndarray, k: int, max_iter: int = 20, seed: int = 42
    ) -> tuple[np.ndarray, np.ndarray]:
        """K-means clustering on binary codes using Hamming distance.

        Args:
            codes: Binary codes, shape (n, n_bytes).
            k: Number of clusters.
            max_iter: Maximum iterations.

        Returns:
            Tuple of (centroids shape (k, n_bytes), assignments shape (n,)).
        """
        n = len(codes)
        rng = np.random.default_rng(seed)

        # Initialize centroids by random selection
        init_indices = rng.choice(n, size=min(k, n), replace=False)
        centroids = codes[init_indices].copy()

        assignments = np.zeros(n, dtype=np.int32)

        for iteration in range(max_iter):
            # Assign each vector to nearest centroid
            new_assignments = np.zeros(n, dtype=np.int32)
            for i in range(n):
                dists = hamming_distance_single(codes[i], centroids)
                new_assignments[i] = int(np.argmin(dists))

            # Check convergence
            if np.array_equal(new_assignments, assignments):
                break
            assignments = new_assignments

            # Update centroids (majority vote per bit position)
            for c in range(k):
                members = codes[assignments == c]
                if len(members) == 0:
                    # Empty cluster: reinitialize randomly
                    centroids[c] = codes[rng.integers(n)]
                else:
                    # Majority vote: for each bit, take the most common value
                    # Unpack to bits, take mean, threshold at 0.5
                    unpacked = np.unpackbits(members, axis=1)
                    majority = (unpacked.mean(axis=0) > 0.5).astype(np.uint8)
                    centroids[c] = np.packbits(majority)

        return centroids, assignments

    def __len__(self) -> int:
        return self._n_vectors

    @property
    def partition_sizes(self) -> list:
        """Number of vectors in each partition."""
        if self._partition_indices is None:
            return []
        return [len(p) for p in self._partition_indices]

    @property
    def vectors_scanned_per_query(self) -> int:
        """Expected number of vectors scanned per query."""
        if self._n_vectors == 0:
            return 0
        avg_partition_size = self._n_vectors / self.n_partitions
        return int(avg_partition_size * self.n_probe)

File: src/partitioned.rs
//! Partition-aware staged retrieval using binary k-means routing.
//!
//! Clusters binary codes into P partitions at build time. At search time,
//! routes the query to the top-R most relevant partitions and scans only those.

use crate::quantize::{normalize, normalize_batch, quantize, quantize_batch};
use crate::search::{hamming_distance, hamming_distance_one_to_many, inner_product};

/// Partition-aware two-stage retrieval with binary k-means routing.
pub struct PartitionedIndex {
    dim: usize,
    n_bytes: usize,
    n_vectors: usize,
    n_partitions: usize,
    n_probe: usize,
    rerank_factor: usize,
    codes: Vec<u8>,           // n * n_bytes
    vectors: Vec<f32>,        // n * dim (unit normalized)
    centroids: Vec<u8>,       // n_partitions * n_bytes
    assignments: Vec<usize>,  // n — partition ID per vector
    partition_indices: Vec<Vec<usize>>, // partition_id -> vector indices
}

impl PartitionedIndex {
    pub fn new(dim: usize, n_partitions: usize, n_probe: usize, rerank_factor: usize) -> Self {
        assert!(dim > 0);
        Self {
            dim,
            n_bytes: (dim + 7) / 8,
            n_vectors: 0,
            n_partitions,
            n_probe,
            rerank_factor,
            codes: Vec::new(),
            vectors: Vec::new(),
            centroids: Vec::new(),
            assignments: Vec::new(),
            partition_indices: Vec::new(),
        }
    }

    /// Build the partitioned index from vectors.
    pub fn build(&mut self, vectors: &[f32]) {
        let n = vectors.len() / self.dim;
        assert_eq!(vectors.len(), n * self.dim);

        self.n_vectors = n;

        // Normalize
        let mut unit_vectors = vectors.to_vec();
        normalize_batch(&mut unit_vectors, n, self.dim);
        self.vectors = unit_vectors.clone();

        // Binary quantize
        self.codes = quantize_batch(&unit_vectors, n, self.dim);

        // Binary k-means clustering
        let (centroids, assignments) = self.binary_kmeans(&self.codes.clone(), n, 20);
        self.centroids = centroids;
        self.assignments = assignments;

        // Build partition index
        self.partition_indices = vec![Vec::new(); self.n_partitions];
        for (i, &pid) in self.assignments.iter().enumerate() {
            self.partition_indices[pid].push(i);
        }
    }

    /// Partition-routed search.
    pub fn search(&self, query: &[f32], k: usize) -> (Vec<f32>, Vec<usize>) {
        if self.n_vectors == 0 {
            return (vec![], vec![]);
        }
        assert_eq!(query.len(), self.dim);

        let mut query_unit = query.to_vec();
        normalize(&mut query_unit);

        let query_code = quantize(&query_unit);

        // Route: find closest partitions by Hamming to centroids
        let centroid_dists = hamming_distance_one_to_many(
            &query_code, &self.centroids, self.n_partitions, self.n_bytes
        );
        let mut partition_order: Vec<(u32, usize)> = centroid_dists.into_iter()
            .enumerate().map(|(i, d)| (d, i)).collect();
        partition_order.sort_unstable_by(|a, b| a.0.cmp(&b.0));

        let n_probe = self.n_probe.min(self.n_partitions);

        // Gather candidates from probed partitions
        let mut candidate_indices: Vec<usize> = Vec::new();
        for &(_, pid) in partition_order.iter().take(n_probe) {
            candidate_indices.extend_from_slice(&self.partition_indices[pid]);
        }

        if candidate_indices.is_empty() {
            return (vec![], vec![]);
        }

        // Binary scan within candidates
        let candidate_codes: Vec<u8> = candidate_indices.iter()
            .flat_map(|&i| self.codes[i * self.n_bytes..(i + 1) * self.n_bytes].iter().copied())
            .collect();
        let hamming_dists = hamming_distance_one_to_many(
            &query_code, &candidate_codes, candidate_indices.len(), self.n_bytes
        );

        // Select top rf*k by Hamming
        let n_rerank = (k * self.rerank_factor).min(candidate_indices.len());
        let mut scored: Vec<(u32, usize)> = hamming_dists.into_iter()
            .enumerate().map(|(local, d)| (d, local)).collect();
        if n_rerank < scored.len() {
            scored.select_nth_unstable_by(n_rerank - 1, |a, b| a.0.cmp(&b.0));
            scored.truncate(n_rerank);
        }

        // Float rerank
        let mut reranked: Vec<(f32, usize)> = scored.iter().map(|&(_, local)| {
            let global = candidate_indices[local];
            let v = &self.vectors[global * self.dim..(global + 1) * self.dim];
            let score = inner_product(&query_unit, v);
            (score, global)
        }).collect();

        reranked.sort_unstable_by(|a, b| b.0.partial_cmp(&a.0).unwrap());
        reranked.truncate(k);

        let scores: Vec<f32> = reranked.iter().map(|(s, _)| *s).collect();
        let indices: Vec<usize> = reranked.iter().map(|(_, i)| *i).collect();
        (scores, indices)
    }

    fn binary_kmeans(&self, codes: &[u8], n: usize, max_iter: usize) -> (Vec<u8>, Vec<usize>) {
        use rand::prelude::*;
        let mut rng = StdRng::seed_from_u64(42);

        let k = self.n_partitions.min(n);

        // Initialize centroids by random selection
        let mut centroid_indices: Vec<usize> = (0..n).collect();
        centroid_indices.shuffle(&mut rng);
        centroid_indices.truncate(k);

        let mut centroids: Vec<u8> = centroid_indices.iter()
            .flat_map(|&i| codes[i * self.n_bytes..(i + 1) * self.n_bytes].iter().copied())
            .collect();

        let mut assignments = vec![0usize; n];

        for _ in 0..max_iter {
            // Assign each vector to nearest centroid
            let mut new_assignments = vec![0usize; n];
            for i in 0..n {
                let code = &codes[i * self.n_bytes..(i + 1) * self.n_bytes];
                let mut best_dist = u32::MAX;
                let mut best_c = 0;
                for c in 0..k {
                    let centroid = &centroids[c * self.n_bytes..(c + 1) * self.n_bytes];
                    let dist = hamming_distance(code, centroid);
                    if dist < best_dist {
                        best_dist = dist;
                        best_c = c;
                    }
                }
                new_assignments[i] = best_c;
            }

            if new_assignments == assignments {
                break;
            }
            assignments = new_assignments;

            // Update centroids: majority vote per bit
            for c in 0..k {
                let members: Vec<usize> = (0..n).filter(|&i| assignments[i] == c).collect();
                if members.is_empty() {
                    // Reinitialize empty cluster
                    let rand_idx = rng.gen_range(0..n);
                    centroids[c * self.n_bytes..(c + 1) * self.n_bytes]
                        .copy_from_slice(&codes[rand_idx * self.n_bytes..(rand_idx + 1) * self.n_bytes]);
                } else {
                    // Majority vote per bit
                    for byte_idx in 0..self.n_bytes {
                        let mut byte_val = 0u8;
                        for bit in 0..8 {
                            let count: usize = members.iter()
                                .filter(|&&m| codes[m * self.n_bytes + byte_idx] & (1 << (7 - bit)) != 0)
                                .count();
                            if count * 2 > members.len() {
                                byte_val |= 1 << (7 - bit);
                            }
                        }
                        centroids[c * self.n_bytes + byte_idx] = byte_val;
                    }
                }
            }
        }

        (centroids, assignments)
    }

    pub fn len(&self) -> usize {
        self.n_vectors
    }

    pub fn is_empty(&self) -> bool {
        self.n_vectors == 0
    }

    pub fn partition_sizes(&self) -> Vec<usize> {
        self.partition_indices.iter().map(|p| p.len()).collect()
    }

    pub fn scan_percentage(&self) -> f64 {
        if self.n_partitions == 0 { return 100.0; }
        (self.n_probe as f64 / self.n_partitions as f64) * 100.0
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_partitioned_build_and_search() {
        let dim = 32;
        let n = 200;
        let vectors: Vec<f32> = (0..dim * n)
            .map(|i| ((i * 7 + 3) % 17) as f32 - 8.0)
            .collect();

        let mut index = PartitionedIndex::new(dim, 8, 4, 50);
        index.build(&vectors);
        assert_eq!(index.len(), n);

        let query = &vectors[0..dim];
        let (scores, indices) = index.search(query, 5);
        assert!(!indices.is_empty());
        assert!(scores[0] > 0.9);
    }
}
